display_list/dl_builder.cc

```diff
@@ -620,7 +620,13 @@ void DisplayListBuilder::RestoreLayer() {
       content_bounds.intersect(layer_op->rect);
     }
   }
-  layer_op->rect = content_bounds;
+  // If the caller specified save layer bounds, preserve those
+  // instead of overwriting with content bounds. The caller
+  // bounds may clip the contents or otherwise limit the bounds
+  // of backdrop entities.
+  if (!layer_op->options.bounds_from_caller()) {
```

Comment:
@flar I think this is a bug in that we can have a saveLayer with bounds but end up blowing away the coverage value. In the case I saw, we did a with bounds and a backdrop filter, the layer itself had no contents. The bounds from caller were overwritten to empty, however the save layer state of bounds from caller was still reported as true, which makes it look like the caller specified empty bounds.

Reply:
In this case I can work around it by ignoring bounds from caller when there is a bdf, as I don't believe its possible to specify saveLayer bounds on backdrop filters from flutter.

Reply:
oh, or maybe its the line above actually

Reply:
Um, the bounds of the *content* of the saveLayer bounds are empty and that is correct. It doesn't matter if the bounds came from the caller or not. Empty saveLayer bounds do not mean that the saveLayer is a NOP, they just mean that there is nothing to draw *into* the bounds.

But, the backdrop filter is a separate aspect of the operation.

Everything is working as intended, but you are interpreting the bounds as if it contains the BDF, but it does not. It is providing you with information about the contents of the layer and you must combine that with other information to figure out the RT size.

Reply:
In this case I am specifying non empty bounds of a particular size on a savayer with no content. The bounds are overwritten to empty here

Reply:
You are specifying non-empty bounds? Are you saying that you have:

```
builder.SaveLayer(100x100 rect, paint, backdrop);
// No render calls here
builder.Restore();
```

if that is the case then you claimed that there was content that was 100x100 in size, but you gave no content, so I corrected it to 0x0. You also specified a backdrop filter which also (completely 100% independently of the saveLayer argument) contribute to the resulting output.

Everything there is working as intended.

If you want to clip the BDF, you have to put a clip around the saveLayer, as in:

```
builder.Save();
builder.ClipRect(100x100 rect);
builder.saveLayer(<doesn't really matter>, paint, bdf);
// No render calls here
builder.Restore();
builder.Restore();
```

(The outer save/restore are only if you are installing that clip just for the SaveLayer.)

Reply:
So bounds from caller really isn't what I wanted then. Bdf aside, the original reason I incorporated it was to try and deal with unbounded contents. But instead of bounds from caller I should be using the content clipped setting then

Thanks for the pointers!

Reply:
> But instead of bounds from caller I should be using the content clipped setting then

Only if there is unbounded content like a BDF or the unbounded flag is set. In all other cases (the very common cases), you should use the bounds. even if BDF is present, some BDFs might not cover the entire clip and so the bounds of the content then becomes relevant.

Note that Skia needs the content bounds, not expanded by anything. They do all of those calculations themselves and don't expect the saveLayer bounds to have already incorporated any of that. If/when DL is only being used for Impeller then  we can revisit how much work it front-loads, but it will never be able to completely front-load everything because of the limitations we've discussed before - not knowing the eventual frame size or the clip/transform inherited from outside a given DL. It could potentially front-load considerations of "the enclosing clip" if it directly saw that clip being recorded, but there are also issues of the full CTM potentially affecting how to interpret that clip which might push that out of reach.

Reply:
I'm not sure what you'd use the content clipped setting for, though. It means that the natural bounds of the content extends outside of the bounds provided in the saveLayer, such as:

```
builder.SaveLayer({10, 10, 20, 20}, null, null);
builder.DrawRect({0, 0, 50, 50}, DlPaint());
builder.Restore();
```

In that case the layer was suggested to capture a 10x10 set of pixels, though the rendering commands draw to a larger area. The legacy behavior of this is that only the 10x10 set of pixels is captured (in scale-translate cases).

The content clipped flag says "warning, the content inside this saveLayer will draw outside the indicated target area". It's primary intent is for the case where group opacity "might" be used because it is a counter-situation. If that SaveLayer had alpha in the paint then you can't just completely elide the SaveLayer like you might do if you are applying opacity inheritance optimizations, because that would draw too much. It has to look like you actually allocated the layer, drew the content, and then faded it - which means some form of clipping would be involved. One solution might be to replace the SaveLayer with a Save and a clip, but so far the Skia implementation has simply avoided the opacity optimization instead. It is possible to notice the difference, but it depends on other factors like transform, etc.

Reply:
A clip will definitely be faster.

> I'm not sure what you'd use the content clipped setting for

basically for destructive blends that flood the clip. But actually it sounds like display list bounds will already take that into account and I can take them as is?